File: jar/src/main/java/jline/api/mam/Qbd_depproc_etaqa_ps.java
/**
 * @file QBD departure process ETAQA truncation for PS discipline
 *
 * @since LINE 3.0
 */
package jline.api.mam;

import java.util.Map;

import jline.lib.smc.QBD_CR;
import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class Qbd_depproc_etaqa_ps {
    private Qbd_depproc_etaqa_ps() {}

    /**
     * Compute MAP departure process for MAP/MAP/1-PS via ETAQA truncation.
     */
    public static MatrixCell qbd_depproc_etaqa_ps(MatrixCell MAPa, MatrixCell MAPs, int n) {
        int na = MAPa.get(0).getNumRows();
        int ns = MAPs.get(0).getNumRows();
        int lvlsz = ns * na;

        Matrix IA = Matrix.eye(na);
        Matrix IS = Matrix.eye(ns);
        Matrix F = MAPa.get(1).kron(IS);
        Matrix L = MAPa.get(0).kron(IS).add(IA.kron(MAPs.get(0)));
        Matrix B = IA.kron(MAPs.get(1));
        Matrix L0 = MAPa.get(0).kron(IS);

        Map<String, Matrix> qbdResult = QBD_CR.QBD_CR(B, L, F, null, null, null, null);
        Matrix R = qbdResult.get("R");
        if (R == null) {
            throw new RuntimeException("QBD_CR failed to compute R matrix");
        }

        Matrix G = L.add(R.mult(B)).neg().inv().mult(B);
        Matrix Lhat = F.add(L);
        Matrix Bbar = B.add(F.mult(G));
        Matrix Bhat = F.mult(G);

        int totalDim = (n + 1) * lvlsz;

        Matrix D0 = Matrix.zeros(totalDim, totalDim);

        D0.insertSubMatrix(0, 0, lvlsz, lvlsz, L0);
        if (n >= 1) {
            D0.insertSubMatrix(0, lvlsz, lvlsz, 2 * lvlsz, F);
        }

        for (int j = 1; j <= n; j++) {
            int rs = j * lvlsz;
            D0.insertSubMatrix(rs, rs, rs + lvlsz, rs + lvlsz, L);
            if (j < n) {
                D0.insertSubMatrix(rs, rs + lvlsz, rs + lvlsz, rs + 2 * lvlsz, F);
            }
        }

        if (n >= 1) {
            int rs = (n - 1) * lvlsz;
            D0.insertSubMatrix(rs, rs, rs + lvlsz, rs + lvlsz, Lhat);
        }

        if (n >= 1) {
            int rsN = n * lvlsz;
            int csN1 = (n - 1) * lvlsz;
            for (int i = 0; i < lvlsz; i++) {
                for (int j = 0; j < lvlsz; j++) {
                    D0.set(rsN + i, csN1 + j, D0.get(rsN + i, csN1 + j) + Bbar.get(i, j));
                }
            }
            double tailScale = (double) (n - 1) / (double) n;
            for (int i = 0; i < lvlsz; i++) {
                for (int j = 0; j < lvlsz; j++) {
                    D0.set(rsN + i, rsN + j, D0.get(rsN + i, rsN + j) + Bhat.get(i, j) * tailScale);
                }
            }
        }

        Matrix D1 = Matrix.zeros(totalDim, totalDim);

        if (n >= 1) {
            int rsN = n * lvlsz;
            int csN1 = (n - 1) * lvlsz;
            D1.insertSubMatrix(rsN, csN1, rsN + lvlsz, csN1 + lvlsz, Bbar);
            D1.insertSubMatrix(rsN, rsN, rsN + lvlsz, rsN + lvlsz, Bhat.scale(1.0 / (double) n));
        }

        for (int j = 1; j < n; j++) {
            int rs = j * lvlsz;
            int cs = (j - 1) * lvlsz;
            double invJ = 1.0 / (double) j;

            double nonDepartScale = 1.0 - invJ;
            for (int i = 0; i < lvlsz; i++) {
                for (int k = 0; k < lvlsz; k++) {
                    D0.set(rs + i, cs + k, D0.get(rs + i, cs + k) + B.get(i, k) * nonDepartScale);
                }
            }

            D1.insertSubMatrix(rs, cs, rs + lvlsz, cs + lvlsz, B.scale(invJ));
        }

        MatrixCell result = new MatrixCell(2);
        result.set(0, D0);
        result.set(1, D1);
        return result;
    }
}

File: matlab/src/api/mam/qbd_depproc_etaqa_ps.m
%{ @file qbd_depproc_etaqa_ps.m
 %  @brief Constructs MAP departure process for MAP/MAP/1-PS via ETAQA truncation
 %
 %  @author LINE Development Team
%}

%{
 % @brief Builds a MAP approximation of the departure process for a MAP/MAP/1 PS queue
 %
 % @details
 % This function constructs a finite-state MAP {D0, D1} representation of the
 % departure process from a MAP/MAP/1 queue with Processor Sharing (PS)
 % discipline, using ETAQA truncation at QBD level n.
 %
 % Compared to the FCFS variant (qbd_depproc_etaqa), the PS discipline splits
 % service completions at level j into a departure component B*(1/j) and an
 % internal transition component B*(1-1/j), reflecting the rate-dependent
 % sharing of the server among j jobs. The tail approximation at level n uses
 % Bbar and Bhat matrices weighted by 1/n and (n-1)/n respectively.
 %
 % @par Syntax:
 % @code
 % D = qbd_depproc_etaqa_ps(MAPa, MAPs, n)
 % @endcode
 %
 % @par Parameters:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>MAPa<td>Arrival process in MAP format {D0, D1}
 % <tr><td>MAPs<td>Service process in MAP format {D0, D1}
 % <tr><td>n<td>Truncation level (number of QBD levels to represent explicitly)
 % </table>
 %
 % @par Returns:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>D<td>Departure process in MAP format {D0, D1}
 % </table>
%}
function [D]=qbd_depproc_etaqa_ps(MAPa,MAPs,n)

na = length(MAPa{1});
ns = length(MAPs{1});
lvlsz = ns*na;

F = kron(MAPa{2},eye(ns));
L = krons(MAPa{1},MAPs{1});
B = kron(eye(na),MAPs{2});
L0 = kron(MAPa{1},eye(ns));

[~,R,~] = QBD_CR(B,L,F);
G = inv(-L-R*B)*B;
Lhat = F+L;
Bbar = B+F*G;
Bhat = F*G;

vn1 = ones(1,n-1);
vn = zeros(1,n); vn(end)=1;

D0=kron(diag([1,vn1]),L)+kron(diag(vn1,1),F);
D0=[zeros(size(L0,1),size(D0,2)); D0];
D0 = [zeros(size(D0,1),size(B,2)),D0];
D0(1:size(L0,1),1:(size(L0,2)+size(F,2)))=[L0,F];
D0(((n-1)*lvlsz+1):n*lvlsz,((n-1)*lvlsz+1):n*lvlsz)=Lhat;

D0 = D0 + kron(diag(vn,-1),Bbar)+kron(diag([0,vn]),Bhat)*(n-1)/n;
D1=kron(diag(vn,-1),Bbar)+kron(diag([0,vn]),Bhat)/n;
for j=1:(n-1)
    D0((j*lvlsz+1):(j+1)*lvlsz, ((j-1)*lvlsz+1):j*lvlsz)=B*(1-1/j);
    D1((j*lvlsz+1):(j+1)*lvlsz, ((j-1)*lvlsz+1):j*lvlsz)=B*(1/j);
end
D={D0,D1};
end
